text/0000-try-trait-v2.md

```diff
@@ -0,0 +1,970 @@
+            Poll::Ready(Err(e)) => ControlFlow::Break(Err(e)),
+            Poll::Pending => ControlFlow::Continue(Poll::Pending),
+        }
+    }
+}
+
+impl<T, E, F: From<E>> ops::FromTryResidual<Result<!, E>> for Poll<Result<T, F>> {
+    fn from_residual(x: Result<!, E>) -> Self {
+        match x {
+            Err(e) => Poll::Ready(Err(From::from(e))),
+        }
+    }
+}
+```
+
+```rust
+impl<T, E> ops::Try for Poll<Option<Result<T, E>>> {
+    type Output = Poll<Option<T>>;
+    type Residual = <Result<T, E> as ops::Try>::Residual;
+
+    fn from_output(c: Self::Output) -> Self {
+        c.map(|x| x.map(Ok))
+    }
+
+    fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
+        match self {
+            Poll::Ready(Some(Ok(x))) => ControlFlow::Continue(Poll::Ready(Some(x))),
+            Poll::Ready(Some(Err(e))) => ControlFlow::Break(Err(e)),
+            Poll::Ready(None) => ControlFlow::Continue(Poll::Ready(None)),
+            Poll::Pending => ControlFlow::Continue(Poll::Pending),
+        }
+    }
+}
+
+impl<T, E, F: From<E>> ops::FromTryResidual<Result<!, E>> for Poll<Option<Result<T, F>>> {
+    fn from_residual(x: Result<!, E>) -> Self {
+        match x {
+            Err(e) => Poll::Ready(Some(Err(From::from(e)))),
+        }
+    }
+}
+```
+
+### `ControlFlow`
+
+```rust
+impl<B, C> ops::Try for ControlFlow<B, C> {
+    type Output = C;
+    type Residual = ControlFlow<B, !>;
+
+    fn from_output(c: C) -> Self {
+        ControlFlow::Continue(c)
+    }
+
+    fn branch(self) -> ControlFlow<Self::Residual, C> {
+        match self {
+            ControlFlow::Continue(c) => ControlFlow::Continue(c),
+            ControlFlow::Break(b) => ControlFlow::Break(ControlFlow::Break(b)),
+        }
+    }
+}
+
+impl<B, C> ops::FromTryResidual for ControlFlow<B, C> {
+    fn from_residual(x: <Self as ops::Try>::Residual) -> Self {
+        match x {
+            ControlFlow::Break(r) => ControlFlow::Break(r),
+        }
+    }
+}
+```
+
+## Making the accidental `Option` interconversion continue to work
+
+This is done with an extra implementation:
+```rust
+mod sadness {
+    use super::*;
+
+    /// This is a remnant of the old `NoneError` which is never going to be stabilized.
+    /// It's here as a snapshot of an oversight that allowed this to work in the past,
+    /// so we're stuck supporting it even though we'd really rather not.
+    #[derive(Clone, Copy, PartialEq, PartialOrd, Eq, Ord, Debug, Hash)]
+    pub struct PleaseCallTheOkOrMethodToUseQuestionMarkOnOptionsInAMethodThatReturnsResult;
+
+    impl<T, E> ops::FromTryResidual<Option<!>> for Result<T, E>
+    where
+        E: From<PleaseCallTheOkOrMethodToUseQuestionMarkOnOptionsInAMethodThatReturnsResult>,
+    {
+        fn from_residual(x: Option<!>) -> Self {
+            match x {
+                None => Err(From::from(
+                    PleaseCallTheOkOrMethodToUseQuestionMarkOnOptionsInAMethodThatReturnsResult,
```

Comment:
My recollection is that there's code actively doing it.  It's relatively easy to do accidentally, as it can be done with just an error type that's `From<impl Debug>`, for example: https://github.com/scottmcm/rust/pull/2/files#diff-896d8e28be5d0f748c0e13bae0a4eadb218503c2768ec5af9633e16af777d5baR5

(Though it's not trivial because of coherence with `T: From<T>`.)

Reply:
About a day left on the crater run to get solid data on this: https://crater.rust-lang.org/ex/pr-82322